File: HyeonjeongKo/day4-155.py
#include <iostream>
#include <string>
#include <stack>


class MinStack {

    int min=0;
    stack<int> s;
    stack<int> m;

public:
    MinStack() {
        min=0;
    }
    
    void push(int val) {
        s.push(val);
        
        if(m.empty()){ // m이 비어있을때
            m.push(val);

        }
        else if(m.top()>=val){ // val이 최소값일때
            m.push(val);
        }
        else{
            int t = m.top();
            m.push(t);
        }
    }
    
    void pop() {
        s.pop();
        m.pop();
    }
    
    int top() {
        return s.top();
    }
    
    int getMin() {
        return m.top();
    }
};

File: JooEun/2023/day4-155.cpp
// 2023.01.05 https://leetcode.com/problems/min-stack/

// #include <algorithm>
// #include <vector>

class MinStack {
public:
    vector<int> stack; // member variable
    
    MinStack() {
      // default constructor
    }
    
    void push(int val) {
        stack.push_back(val);
    }
    
    void pop() {
        stack.pop_back();
    }
    
    int top() {
        return stack.back();
    }
    
    int getMin() {
        // https://notepad96.tistory.com/40 :: use algorithm library
        return *min_element(stack.begin(), stack.end());
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */
